js-api-doc/importer.d.ts

```diff
@@ -3,32 +3,293 @@ import {URL} from 'url';
+ *
+ * - The compiler encounters `@use "db:foo/bar/baz"`.
+ * - It calls [[canonicalize]] with `"db:foo/bar/baz"`.
+ * - [[canonicalize]] returns `new URL("db:foo/bar/baz/_index.scss")`.
+ * - If the compiler has already loaded a stylesheet with this canonical URL, it
+ *   re-uses the existing module.
+ * - Otherwise, it calls [[load]] with `new URL("db:foo/bar/baz/_index.scss")`.
+ * - [[load]] returns an [[ImporterResult]] that the compiler uses as the
+ *   contents of the module.
+ *
+ * @example Code Sample
+ *
+ * ```js
+ * sass.compile('style.scss', {
+ *   // An importer for URLs like `bgcolor:orange` that generates a
+ *   // stylesheet with the given background color.
+ *   importers: [{
+ *     canonicalize(url) {
+ *       if (!url.startsWith('bgcolor:')) return null;
+ *       return new URL(url);
+ *     },
+ *     load(canonicalUrl) {
+ *       return {
+ *         contents: `body {background-color: ${canonicalUrl.pathname}}`,
+ *         syntax: 'scss'
+ *       };
+ *     }
+ *   }]
+ * });
+ * ```
+ *
+ * @category Importer
+ */
 export interface Importer<sync extends 'sync' | 'async' = 'sync' | 'async'> {
+  /**
+   * If `url` is recognized by this importer, returns its canonical format.
+   *
+   * If Sass has already loaded a stylesheet with the returned canonical URL, it
+   * re-uses the existing parse tree (and the loaded module for `@use`). This
+   * means that importers **must ensure** that the same canonical URL always
+   * refers to the same stylesheet, *even across different importers*. As such,
+   * importers are encouraged to use unique URL schemes to disambiguate between
+   * one another.
+   *
+   * As much as possible, custom importers should canonicalize URLs the same way
+   * as the built-in filesystem importer:
+   *
+   * - The importer should look for stylesheets by adding the prefix `_` to the
+   *   URL's basename, and by adding the extensions `.sass` and `.scss` if the
+   *   URL doesn't already have one of those extensions. For example, if the
+   *   URL was `foo/bar/baz`, the importer would look for:
+   *   - `foo/bar/baz.sass`
+   *   - `foo/bar/baz.scss`
+   *   - `foo/bar/_baz.sass`
+   *   - `foo/bar/_baz.scss`
+   *
+   *   If the URL was `foo/bar/baz.scss`, the importer would just look for:
+   *   - `foo/bar/baz.scss`
+   *   - `foo/bar/_baz.scss`
+   *
+   *   If the importer finds a stylesheet at more than one of these URLs, it
+   *   should throw an exception indicating that the URL is ambiguous. Note that
+   *   if the extension is explicitly specified, a stylesheet with the opposite
+   *   extension is allowed to exist.
+   *
+   * - If none of the possible paths is valid, the importer should perform the
+   *   same resolution on the URL followed by `/index`. In the example above,
+   *   it would look for:
+   *   - `foo/bar/baz/_index.sass`
+   *   - `foo/bar/baz/index.sass`
+   *   - `foo/bar/baz/_index.scss`
+   *   - `foo/bar/baz/index.scss`
```

Comment:
Nit: Is order meaningful here? If not consistency with previous example could be nice:

```suggestion
   *   - `foo/bar/baz/index.sass`
   *   - `foo/bar/baz/index.scss`
   *   - `foo/bar/baz/_index.sass`
   *   - `foo/bar/baz/_index.scss`
```